### PR TITLE
Optimizations, idiomatic changes

### PR DESCRIPTION
**Pull request by nascent-chris**
This PR is not intended to add/remove/change any functionality, only to help improve what exists, e.g. by
* reducing data copying (borrowing where possible, removing `.clone()` where possible, `String` -> `&str`, etc.)
* removing `.unwrap()` calls wherever possible
* reducing need for `mut` variables

Some changes here may be controversial, such as introducing `anyhow` as a dependency.  Interested in hearing arguments for/against.

### CODE REVIEW

File: src/ws/mod.rs [hunk 1]
```diff
@@ -90,12 +87,12 @@ impl Ws {
 
     /// Subscribe to specified `Channel`s
     /// For FILLS the socket needs to be authenticated
-    pub async fn subscribe(&mut self, channels: Vec<Channel>) -> Result<()> {
+    pub async fn subscribe(&mut self, channels: &[Channel]) -> Result<()> {
```
Comment (dovahcrow) [inline src/ws/mod.rs:90]: Would it be better to have channels be a generic `I: IntoIter<Channle>`

**Review by MaxFangX — APPROVED** [review]
This was a much needed cleaning and optimization, thank you.

I went ahead and tried it out in my own project - there are a ton of API breaking changes (every `String` -> `'a str`), but given that this is a trading library where performance is quite important, it is worth it IMO. Existing users may need to redefine their own "owned" versions of a request struct if they are sending this struct across threads, which can be fairly easily done by importing the `Request` trait and copying over the struct, changing `str` back to `String`. After merge we should bump the version to 0.6, and maybe document how users can customize this crate to suit their needs.

I would like to push back on changing the Result type to `anyhow::Result`. I love `anyhow`, and I'm using anyhow extensively in my own project that depends on this library, but I also believe that anyhow should be reserved for applications; libraries like this should expose clean Error enums so that consumers can match on the contained errors. This isn't just hypothetical - I have a decent amount of logic that matches on the `Error` enums in this crate to differentiate between recoverable network errors that need a reconnect (e.g. `tungstenite::error::ProtocolError::ResetWithoutClosingHandshake`), errors from simple race conditions that don't need any special handling (e.g. `tungstenite::Error::AlreadyClosed`), and others. Using `anyhow` not only makes this more difficult, but also more brittle - I'd have to base my classification on whether the anyhow error strings contain a particular substring, which breaks whenever any change is made to the error string, instead of producing a compilation error whenever some API change was made in the underlying library.

Overall I'm looking forward to using these optimizations!

**Comment by nascent-chris** [comment]
> I would like to push back on changing the Result type to `anyhow::Result`. I love `anyhow`, and I'm using anyhow extensively in my own project that depends on this library, but I also believe that anyhow should be reserved for applications; libraries like this should expose clean Error enums so that consumers can match on the contained errors. This isn't just hypothetical - I have a decent amount of logic that matches on the `Error` enums in this crate to differentiate between recoverable network errors that need a reconnect (e.g. `tungstenite::error::ProtocolError::ResetWithoutClosingHandshake`), errors from simple race conditions that don't need any special handling (e.g. `tungstenite::Error::AlreadyClosed`), and others. Using `anyhow` not only makes this more difficult, but also more brittle - I'd have to base my classification on whether the anyhow error strings contain a particular substring, which breaks whenever any change is made to the error string, instead of producing a compilation error whenever some API change was made in the underlying library.

Yeah you've convinced me as well.  I'll refactor to remove it.  Thank you for the feedback!

File: src/ws/tests.rs [hunk 2]
```diff
@@ -17,31 +17,31 @@ async fn init_unauthenticated_ws() -> Ws {
...
 
 #[allow(dead_code)]
-async fn init_api() -> Rest {
-    dotenv().ok();
+async fn init_api() -> Result<Rest> {
+    dotenv().expect("dotenv panic in init_api()");
 
-    Rest::new(Options::from_env())
+    Ok(Rest::new(Options::from_env()))
 }
```
Comment (MaxFangX) [inline src/ws/tests.rs:24]: `Result` no longer needed here

File: src/rest/tests.rs [hunk 3]
```diff
@@ -22,11 +22,13 @@ async fn init_unauthenticated_api() -> Rest {
...
-        Err(Error::Api(error)) if error == *"Not allowed with read-only permissions" => {}
-        _ => panic!("Expected read-only subaccount."),
-    }
+fn read_only<T: std::fmt::Debug>(result: Result<T>) {
+    result.expect_err("Expected error");
+    //TODO: match specific error
+    // match result {
+    //     Err(Error::Api(error)) if error == *"Not allowed with read-only permissions" => {}
+    //     _ => panic!("Expected read-only subaccount."),
+    // }
```
Comment (MaxFangX) [inline src/rest/tests.rs:31]: Looks like this can be uncommented again?

**Review by MaxFangX — APPROVED** [review]
Still looks good, just some nits in the tests.

I also noticed that updating the orderbook no longer checks whether `OrderbookAction` is `Partial` - the [docs](https://docs.ftx.com/#orderbooks) seem to indicate that a `Partial` update is only received once, so I guess this is fine?

I'd merge but since this is a larger change I want to make sure @fabianboesiger has had a chance to provide some input first.

**Comment by fabianboesiger** [comment]
Hi, thanks for the PR! The use of `String` and `.clone()` was mostly because I quickly needed a working library, so your changes to use `&str` are more than welcome! I agree with @MaxFangX that using anyhow isn't necessary, but it seems like you already have removed it as a dependency.

**Pull request merged**